In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [3]:
df_sold = pd.read_csv("sold_residential.csv")
df_list = pd.read_csv("listing_residential.csv")

C:\Users\junki\AppData\Local\Temp\ipykernel_16376\2745887180.py:1: DtypeWarning: Columns (0: BuyerAgencyCompensationType) have mixed types. Specify dtype option on import or set low_memory=False.
  df_sold = pd.read_csv("sold_residential.csv")
C:\Users\junki\AppData\Local\Temp\ipykernel_16376\2745887180.py:2: DtypeWarning: Columns (0: BuyerAgencyCompensationType) have mixed types. Specify dtype option on import or set low_memory=False.
  df_list = pd.read_csv("listing_residential.csv")


In [4]:
print(df_sold.shape)
df_sold.head()

(430075, 61)


,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,CloseDate,ClosePrice,ListAgentFirstName,...,HighSchool,Levels,NewConstructionYN,GarageSpaces,HighSchoolDistrict,PostalCode,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,BuyerAgencyCompensationType,BuyerAgencyCompensation
0,NaN,True,NaN,NaN,True,598000.0,1048490409,2024-01-01,555000.0,Shelli,...,NaN,One,False,2.0,Riverside Unified,92504,6969.0,NaN,NaN,NaN
1,Laminate,False,NaN,NaN,False,915000.0,1046722825,2024-01-01,915000.0,Esperanza,...,NaN,Two,NaN,2.0,Moorpark Unified,93021,4649.0,NaN,NaN,NaN
2,Wood,True,NaN,NaN,False,184500.0,1048590625,2024-01-01,186500.0,Nathan,...,NaN,One,False,0.0,NaN,92026,581115.0,NaN,NaN,NaN
3,NaN,True,NaN,NaN,False,1600000.0,1054052014,2024-01-01,1600000.0,Joseph,...,NaN,Two,False,2.0,NaN,92054,13567.0,NaN,NaN,NaN
4,"Laminate,Tile",False,NaN,NaN,False,535000.0,1046369026,2024-01-01,535000.0,Gustavo,...,NaN,One,False,2.0,Rialto Unified,92376,7700.0,NaN,NaN,NaN


In [5]:
pd.set_option("display.max_rows", None)

In [6]:
# check all features for percentage of missing values
missing_ratio = df_sold.isna().sum() / df_sold.shape[0]
print(missing_ratio.sort_values(ascending=False))

# call features that contain more than 90% of missing values
missing_high = missing_ratio[missing_ratio>=0.9]
print()
print("Missing more than 90%")
print(missing_high.sort_values(ascending=False))

TaxAnnualAmount                 1.000000
FireplacesTotal                 1.000000
MiddleOrJuniorSchoolDistrict    1.000000
TaxYear                         1.000000
ElementarySchoolDistrict        1.000000
BusinessType                    1.000000
WaterfrontYN                    0.999370
BasementYN                      0.980399
BuilderName                     0.951041
BuyerAgencyCompensation         0.892802
BuyerAgencyCompensationType     0.892777
ElementarySchool                0.866986
MiddleOrJuniorSchool            0.866100
HighSchool                      0.826095
SubdivisionName                 0.627509
Flooring                        0.358775
HighSchoolDistrict              0.275336
Stories                         0.153648
AttachedGarageYN                0.152792
MLSAreaMajor                    0.134191
Levels                          0.096002
PoolPrivateYN                   0.085897
ViewYN                          0.085504
LotSizeSquareFeet               0.077221
NewConstructionY

In [7]:
# drop features over 90% missing values
df_sold = df_sold.drop(columns = missing_high.index)

(df_sold.isna().sum() / df_sold.shape[0]) >= 0.9

Flooring                       False
ViewYN                         False
PoolPrivateYN                  False
OriginalListPrice              False
ListingKey                     False
CloseDate                      False
ClosePrice                     False
ListAgentFirstName             False
ListAgentLastName              False
Latitude                       False
Longitude                      False
UnparsedAddress                False
PropertyType                   False
LivingArea                     False
ListPrice                      False
DaysOnMarket                   False
ListOfficeName                 False
BuyerOfficeName                False
ListAgentFullName              False
BuyerAgentMlsId                False
BuyerAgentFirstName            False
BuyerAgentLastName             False
ListingKeyNumeric              False
MLSAreaMajor                   False
CountyOrParish                 False
MlsStatus                      False
ElementarySchool               False
A

In [8]:
# call features have missing values between 60% to 90%
missing_mid = missing_ratio[(missing_ratio>=0.6)&(missing_ratio<0.9)]
missing_mid.sort_values(ascending=False)

BuyerAgencyCompensation        0.892802
BuyerAgencyCompensationType    0.892777
ElementarySchool               0.866986
MiddleOrJuniorSchool           0.866100
HighSchool                     0.826095
SubdivisionName                0.627509
dtype: float64

These are the features that are not consistently reported across the dataset, because more than half of records are missing. For this analysis, we identified as high missing features 60% to 90% that we avoid using them for major conclusions unless we specifically focus on the subset of records.

In [9]:
# call features have missing values between 30% to 60%
missing_mid_low = missing_ratio[(missing_ratio>=0.3)&(missing_ratio<0.6)]
missing_mid_low.sort_values(ascending=False)

Flooring    0.358775
dtype: float64

Flooring feature had moderate missing values of 35.88% that one-third of records did not report flooring information, and we should interpret with caution.


In [10]:
df_sold.dtypes

Flooring                           str
ViewYN                          object
PoolPrivateYN                   object
OriginalListPrice              float64
ListingKey                       int64
CloseDate                          str
ClosePrice                     float64
ListAgentFirstName                 str
ListAgentLastName                  str
Latitude                       float64
Longitude                      float64
UnparsedAddress                    str
PropertyType                       str
LivingArea                     float64
ListPrice                      float64
DaysOnMarket                     int64
ListOfficeName                     str
BuyerOfficeName                    str
ListAgentFullName                  str
BuyerAgentMlsId                    str
BuyerAgentFirstName                str
BuyerAgentLastName                 str
ListingKeyNumeric                int64
MLSAreaMajor                       str
CountyOrParish                     str
MlsStatus                

In [11]:
numeric = [
    'OriginalListPrice',
    'ClosePrice',
    'Latitude',
    'Longitude',
    'LivingArea',
    'ListPrice',
    'ParkingTotal',
    'YearBuilt',
    'BathroomsTotalInteger',
    'Stories',
    'GarageSpaces',
    'LotSizeSquareFeet',
    'BuyerAgencyCompensation'
]

discrete = [
    'DaysOnMarket',
    'BedroomsTotal'
]

categorical = [
    'Flooring',
    'ViewYN',
    'PoolPrivateYN',
    'ListAgentFirstName',
    'ListAgentLastName',
    'UnparsedAddress',
    'PropertyType',
    'ListOfficeName',
    'BuyerOfficeName',
    'ListAgentFullName',
    'BuyerAgentMlsId',
    'BuyerAgentFirstName',
    'BuyerAgentLastName',
    'MLSAreaMajor',
    'CountyOrParish',
    'MlsStatus',
    'ElementarySchool',
    'AttachedGarageYN',
    'PropertySubType',
    'SubdivisionName',
    'City',
    'StateOrProvince',
    'MiddleOrJuniorSchool',
    'FireplaceYN',
    'HighSchool',
    'Levels',
    'NewConstructionYN',
    'HighSchoolDistrict',
    'PostalCode',
    'BuyerAgencyCompensationType'
]

date_feature = [
    'CloseDate',
    'ContractStatusChangeDate',
    'PurchaseContractDate',
    'ListingContractDate'
]

id_feature = [
    'ListingKey',
    'ListingKeyNumeric',
    'ListingId'
]

In [12]:
df_sold[numeric+discrete].dtypes

OriginalListPrice          float64
ClosePrice                 float64
Latitude                   float64
Longitude                  float64
LivingArea                 float64
ListPrice                  float64
ParkingTotal               float64
YearBuilt                  float64
BathroomsTotalInteger      float64
Stories                    float64
GarageSpaces               float64
LotSizeSquareFeet          float64
BuyerAgencyCompensation    float64
DaysOnMarket                 int64
BedroomsTotal              float64
dtype: object

In [13]:
df_sold[numeric+discrete].isna().sum()/df_sold.shape[0]

OriginalListPrice          0.001835
ClosePrice                 0.000002
Latitude                   0.037389
Longitude                  0.037389
LivingArea                 0.000570
ListPrice                  0.000000
ParkingTotal               0.000991
YearBuilt                  0.000888
BathroomsTotalInteger      0.000163
Stories                    0.153648
GarageSpaces               0.042983
LotSizeSquareFeet          0.077221
BuyerAgencyCompensation    0.892802
DaysOnMarket               0.000000
BedroomsTotal              0.000028
dtype: float64

In [14]:
df_sold[date_feature].dtypes

CloseDate                   str
ContractStatusChangeDate    str
PurchaseContractDate        str
ListingContractDate         str
dtype: object

In [15]:
for i in date_feature:
    print(f"{i} feature of values")
    print(df_sold[i].unique().tolist())
    print()

CloseDate feature of values
['2024-01-01', '2024-01-02', '2024-01-03', '2024-01-04', '2024-01-05', '2024-01-06', '2024-01-08', '2024-01-09', '2024-01-10', '2024-01-11', '2024-01-12', '2024-01-13', '2024-01-14', '2024-01-15', '2024-01-16', '2024-01-17', '2024-01-18', '2024-01-19', '2024-01-20', '2024-01-21', '2024-01-22', '2024-01-23', '2024-01-24', '2024-01-25', '2024-01-26', '2024-01-27', '2024-01-28', '2024-01-29', '2024-01-30', '2024-01-31', '2024-02-01', '2024-02-02', '2024-02-03', '2024-02-04', '2024-02-05', '2024-02-06', '2024-02-07', '2024-02-08', '2024-02-09', '2024-02-10', '2024-02-11', '2024-02-12', '2024-02-13', '2024-02-14', '2024-02-15', '2024-02-16', '2024-02-17', '2024-02-18', '2024-02-19', '2024-02-20', '2024-02-21', '2024-02-22', '2024-02-23', '2024-02-24', '2024-02-25', '2024-02-26', '2024-02-27', '2024-02-28', '2024-02-29', '2024-03-01', '2024-03-02', '2024-03-03', '2024-03-04', '2024-03-05', '2024-03-06', '2024-03-07', '2024-03-08', '2024-03-09', '2024-03-10', '2024

In [16]:
df_sold[date_feature].isna().sum()/df_sold.shape[0]

CloseDate                   0.000000
ContractStatusChangeDate    0.001370
PurchaseContractDate        0.000458
ListingContractDate         0.000002
dtype: float64

In [17]:
df_sold[id_feature].dtypes

ListingKey           int64
ListingKeyNumeric    int64
ListingId              str
dtype: object

In [18]:
df_sold["ListingId"].unique().tolist()

['CV23213916',
 '223004385',
 '230022940SD',
 '240000171SD',
 'PW23191300',
 'SR23171344',
 'ND23163499',
 'SR24000121',
 '230018444SD',
 'OC23136962',
 'OC23212809',
 'SW23098126',
 'SR23197356',
 'NP23171166',
 'PW23227356',
 'OC23211799',
 '41042711',
 'SW23197702',
 'SW23197763',
 'LC23099722',
 '23322031',
 'NDP2308263',
 '230020766SD',
 'NP23101006',
 '23314501',
 '230022668SD',
 '41044666',
 'LC23170454',
 'NDP2308242',
 'NDP2308244',
 '219101723DA',
 'PTP2305710',
 'OC23196711',
 '219101729DA',
 'OC23194847',
 '230024137SD',
 '23332943',
 'OC23146470',
 'NP23212959',
 'ML81945889',
 'ML81948089',
 'SW23212530',
 'TR23103090',
 'ML81931755',
 'SW23106567',
 'OC23199318',
 'PTP2305253',
 'DW23211940',
 'PW23211895',
 '41042881',
 '219102940DA',
 'PW23210583',
 'ML81947976',
 'ML81943023',
 'NDP2306150',
 'SC23209844',
 'SC23211511',
 '41044460',
 'SR23197421',
 'NDP2308284',
 'NDP2308856',
 'SR23194332',
 'ND23196480',
 'SR23179114',
 '230022516SD',
 'ML81948059',
 '230022508SD',

In [19]:
df_sold[id_feature].isna().sum()/df_sold.shape[0]

ListingKey           0.0
ListingKeyNumeric    0.0
ListingId            0.0
dtype: float64

In [20]:
df_sold[categorical].dtypes

Flooring                          str
ViewYN                         object
PoolPrivateYN                  object
ListAgentFirstName                str
ListAgentLastName                 str
UnparsedAddress                   str
PropertyType                      str
ListOfficeName                    str
BuyerOfficeName                   str
ListAgentFullName                 str
BuyerAgentMlsId                   str
BuyerAgentFirstName               str
BuyerAgentLastName                str
MLSAreaMajor                      str
CountyOrParish                    str
MlsStatus                         str
ElementarySchool                  str
AttachedGarageYN               object
PropertySubType                   str
SubdivisionName                   str
City                              str
StateOrProvince                   str
MiddleOrJuniorSchool              str
FireplaceYN                    object
HighSchool                        str
Levels                            str
NewConstruct

In [27]:
implicit_missing_values = [
    '',
    ' ',
    'Unknown',
    'unknown',
    'UNKNOWN',
    'N/A',
    'NA',
    'na',
    'None',
    'none',
    'NULL',
    'null',
    '--',
    '-',
    'Missing',
    'missing'
]

for i in categorical:
    col = df_sold[i]
    implicit = col[col.isin(implicit_missing_values)].unique().tolist()
    print(f"{i} feature of values")
    print(implicit)
    print()

Flooring feature of values
[]

ViewYN feature of values
[]

PoolPrivateYN feature of values
[]

ListAgentFirstName feature of values
['-']

ListAgentLastName feature of values
[]

UnparsedAddress feature of values
[]

PropertyType feature of values
[]

ListOfficeName feature of values
[]

BuyerOfficeName feature of values
['UNKNOWN']

ListAgentFullName feature of values
[]

BuyerAgentMlsId feature of values
['none']

BuyerAgentFirstName feature of values
['Unknown', '-', 'none']

BuyerAgentLastName feature of values
['UNKNOWN', '-', 'none']

MLSAreaMajor feature of values
[]

CountyOrParish feature of values
[]

MlsStatus feature of values
[]

ElementarySchool feature of values
['Unknown']

AttachedGarageYN feature of values
[]

PropertySubType feature of values
[]

SubdivisionName feature of values
['none', 'Unknown', 'unknown', 'UNKNOWN', '-', 'na']

City feature of values
['Unknown']

StateOrProvince feature of values
[]

MiddleOrJuniorSchool feature of values
[]

FireplaceYN featur

In [ ]:
df_sold["ClosePrice"].describe()